From 7f6e7850cb9ae2b51781fd0bf117753774b601d2 Mon Sep 17 00:00:00 2001
From: Jiang Wu <wujiang@kylinos.cn>
Date: Wed, 27 May 2026 14:29:05 +0800
Subject: [PATCH 15/20] fix: add llama-cli supprot long prompt
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add llama-cli supprot long context test.

Bug: #572259

From: kylin

Severity: Important

Changelog-Bug: #572259 无法测试长文本优化功能
Change-Id: Idd848eba415b4d08f547e61a40f984b1ae636beb
---
 debian/changelog       | 9 +++++++++
 examples/main/main.cpp | 9 ++++++---
 2 files changed, 15 insertions(+), 3 deletions(-)

diff --git a/debian/changelog b/debian/changelog
index 5d44ec6..358404a 100644
--- a/debian/changelog
+++ b/debian/changelog
@@ -1,3 +1,12 @@
+kytensor-llm (2.0.0-ok19k1.0update4) v11-2603-preview; urgency=medium
+
+  * BUG: #572259 无法测试长文本优化功能
+  * 需求: 无
+  * 其他改动说明: 无
+  * 其他改动影响域: 无
+
+ -- liuyihu <liuyihu@kylinos.cn>  Sun, 1 Jun 2026 15:08:34 +0800
+
 kytensor-llm (2.0.0-ok19k1.0update3) v11-2603-preview; urgency=medium
 
   * BUG: 无
diff --git a/examples/main/main.cpp b/examples/main/main.cpp
index 7b5bf01..1133e91 100644
--- a/examples/main/main.cpp
+++ b/examples/main/main.cpp
@@ -350,10 +350,13 @@ int main(int argc, char ** argv) {
         }
     }
 
-    // Tokenize negative prompt
     if ((int) embd_inp.size() > n_ctx - 4) {
-        LOG_ERR("%s: prompt is too long (%d tokens, max %d)\n", __func__, (int) embd_inp.size(), n_ctx - 4);
-        return 1;
+        if (llama_kv_self_expansion(ctx, embd_inp.size())) {
+            n_ctx = llama_n_ctx(ctx);
+        } else {
+            LOG_ERR("%s: prompt is too long (%d tokens, max %d)\n", __func__, (int) embd_inp.size(), n_ctx - 4);
+            return 1;
+        }
     }
 
     // debug message about similarity of saved session, if applicable
-- 
2.43.0

